### ISSUE TITLE
[Platform] Make the health check email configuration overridable from platform

### ISSUE THREAD

**Issue by ramkumarvs** [yugabyte/yugabyte-db#4170]
Right now we don't support overriding the health check out bout email configuration from platform.

Aha! Link: https://yugabyte-test.aha.io/features/PLATFORM-529

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-4170-analysis.md`):

# Analysis: yugabyte/yugabyte-db#4170 — [Platform] Make the health check email configuration overridable from platform

- **Issue**: https://github.com/yugabyte/yugabyte-db/issues/4170 (OPEN, filed 2020-04-08, area/platform, zero comments; Aha PLATFORM-529)
- **Kind**: feature request → behavior-delta validation (current capability vs requested), per protocol Step 1
- **Analysis date**: 2026-08-31
- **Analyst checkout**: master @ `46a9f7adfd226d7bc3315c38b660eef69bb11196` (2026-08-28)

## Verdict

NOT-REPRODUCED — the premise ("Right now we don't support overriding the health check outbound email configuration from platform") is obsolete: the exact capability shipped on 2020-04-15/16 via sibling issue #4201 (commit `f119ba52cbdd`, Phorge D8309) with the UI following via #4443 (commit `149442596529`, 2020-05-26). #4170 was never back-referenced or closed. Quick verdict mapping: **Fixed** (recommend close as completed / duplicate of delivered #4201).

## Root cause

Behavior-delta, so "root cause" = the filing-time gap plus the current implementation that closes it.

**Era gap (filing time, ref `33fc488d877e5cd30f2fb52ffd7715fc864a1426` = parent of the fix, 2020-04-15):**
- Era `managed/src/main/java/com/yugabyte/yw/common/HealthManager.java:73-84` (evidence copy: `/tmp/claude/yb4170-era-HealthManager.java.txt`) — outbound email settings came exclusively from yugaware `appConfig` keys `yb.health.ses_email_username`, `yb.health.ses_email_password`, `yb.health.default_email`, exported as `YB_ALERTS_USERNAME/PASSWORD/EMAIL` env vars to `bin/cluster_health.py`, which sent the mail. Changing the outbound SMTP identity required editing yugaware's `application.conf` and restarting — no platform API/UI surface existed. That is exactly the "outbound email configuration" ("out bout" in the issue body) the reporter wanted overridable.
- Era `HealthChecker.java:358-367` (evidence copy: `/tmp/claude/yb4170-era-HealthChecker.java.txt`) — destination recipients (`alertingData.alertingEmail`, `sendAlertsToYb`) were *already* per-customer platform-settable at filing time; the gap was specifically the outbound (SMTP) side.

**Current implementation (master `46a9f7adfd22`):**
- `managed/src/main/java/com/yugabyte/yw/controllers/CustomerController.java:212-283` — `PUT /customers/:cUUID` accepts `smtpData` (create at :272-274, edit at :275-277, reset-to-default via `smtpData: null` at :279-283) and `alertingData` (:215-269).
- `managed/src/main/java/com/yugabyte/yw/models/configs/CustomerConfig.java:265-266,309-310` — per-customer SMTP config persisted as `ConfigType.ALERTS` / name `"smtp info"`.
- `managed/src/main/java/com/yugabyte/yw/common/EmailHelper.java:254-266` — `getSmtpData(customerUUID)` reads the customer SMTP config; `:225-241` `getDestinations(customerUUID)`; `:135-145` empty fields fall back to runtime-config defaults `yb.health.default_smtp_server` / `default_smtp_port[_ssl]`.
- `managed/src/main/java/com/yugabyte/yw/commissioner/HealthChecker.java:402,421` — the health-check email report is sent with the per-customer SMTP data (`sendEmailReport`, :395-430) to per-customer destinations (:681).
- UI: `managed/ui/src/components/profile/AlertProfileForm.jsx:246` (Alert emails), `:269-331` (SMTP server/port/from/username/password/SSL/TLS), `:168` (`smtpData = null` reverts to defaults).
- Beyond the ask: per-alert-channel SMTP override — `managed/src/main/java/com/yugabyte/yw/common/alerts/impl/AlertChannelEmail.java:52-53`, `managed/ui/src/components/alerts/AlertConfiguration/AddDestinationChannelForm.jsx:521-597`.
- Docs: `docs/content/stable/yugabyte-platform/alerts-monitoring/set-up-alerts-health-check.md:164-170` (Admin > Alert Configurations > Health: email notifications, Email From, SMTP server details).

**Master still affected?** No — capability present end-to-end (API, model, health-check consumer, UI, docs).

## Trigger matrix

Behavior-delta matrix: capability predicted from the issue text vs observed in source at each ref.

| Capability | Filing era (ref `33fc488d877e`) — predicted | Filing era — observed | Master `46a9f7adfd22` — predicted | Master — observed |
|---|---|---|---|---|
| Health-check email recipients settable from platform | settable (not the complaint) | SETTABLE — era HealthChecker.java:358-367 (`alertingData.alertingEmail`, `sendAlertsToYb`) | settable | SETTABLE — EmailHelper.java:225-241; AlertProfileForm.jsx:246 |
| Outbound SMTP server/port settable from platform | NOT settable (the gap) | NOT SETTABLE — no API/UI; only `yb.health.*` appConfig (era HealthManager.java:73-84) | settable | SETTABLE — CustomerController.java:271-277; AlertProfileForm.jsx:277-291 |
| Outbound SMTP credentials settable from platform | NOT settable (the gap) | NOT SETTABLE — `yb.health.ses_email_username/password` appConfig only | settable | SETTABLE — CustomerConfig.java:265-266; AlertProfileForm.jsx:293-307; optional creds per #4952 |
| Email From address settable from platform | NOT settable (the gap) | NOT SETTABLE — fixed `yb.health.default_email` → `YB_ALERTS_EMAIL` | settable | SETTABLE — `smtpData.emailFrom` AlertProfileForm.jsx:269; fallback EmailHelper.java:262-264 |
| Revert override to shipped defaults | n/a | n/a | supported | SUPPORTED — CustomerController.java:279-283 (`smtpData: null` deletes config); EmailHelper.java:135-145 runtime defaults |

All predictions matched observation; the two-sided matrix (recipients already settable, outbound not) isolates the delta to exactly what #4201 delivered.

## Reproduction

Executable reproduction does not apply: this is a YBA control-plane feature ask, not a DB defect — there is no failure signature to trigger, and a `yugabytedb/yugabyte` DB container does not run YBA. Validation evidence is source-state comparison at pinned refs, reproducible by any maintainer:

```sh
# Era gap (parent of the fix, 2020-04-15): outbound email = appConfig-only
gh api "repos/yugabyte/yugabyte-db/contents/managed/src/main/java/com/yugabyte/yw/common/HealthManager.java?ref=33fc488d877e5cd30f2fb52ffd7715fc864a1426" \
  --jq '.content' | base64 -d | grep -n "yb.health"
#  → yb.health.ses_email_username / ses_email_password / default_email (lines 73-84); no smtpData anywhere

# Delivery commit (closed sibling #4201, 8 days after #4170 was filed)
gh api repos/yugabyte/yugabyte-db/commits/f119ba52cbdd --jq '.commit.message' | head -5
#  → "[#4201]: Allow customization of SMTP server for sending alert mails."
#    (adds smtpData customer-config entry; test plan shows PUT /api/v1/customers/:cUUID with smtpData payload)

# Sibling issues closed
gh issue view 4201 --repo yugabyte/yugabyte-db --json state,closedAt   # CLOSED 2020-04-16
gh issue view 4443 --repo yugabyte/yugabyte-db --json state,closedAt   # CLOSED 2020-06-02 (UI)

# Current master: per-customer SMTP consumed by the health check
grep -n "getSmtpData" managed/src/main/java/com/yugabyte/yw/commissioner/HealthChecker.java   # :402
grep -n "createSmtpConfig\|getSmtpConfig" managed/src/main/java/com/yugabyte/yw/models/configs/CustomerConfig.java  # :265,:309
```

Era file snapshots preserved at `/tmp/claude/yb4170-era-HealthManager.java.txt` and `/tmp/claude/yb4170-era-HealthChecker.java.txt` (byte-identical `.txt`-suffixed copies of the GitHub contents-API downloads at ref `33fc488d877e`).

## Versions tested

- master @ `46a9f7adfd226d7bc3315c38b660eef69bb11196` (this checkout, 2026-08-28): capability present. Master affected: **no**.
- Era ref `33fc488d877e5cd30f2fb52ffd7715fc864a1426` (2020-04-15, parent of fix ≈ filing time): gap present (via GitHub contents API; local clone is shallow).
- Delivery commits: `f119ba52cbdd` (2020-04-15, #4201/D8309 — API + backend), `149442596529` (2020-05-26, #4443 — UI). Hardening follow-ups: #4952 (optional SMTP creds), #7698/#7703/#7810 (masking, no-auth, default-config fixes), PLAT-11580 (custom CA trust store for SMTP, `EmailHelper.java:165-185`).

## Proposed fix direction

No code fix warranted — the capability shipped in 2020 and has since been extended (per-channel SMTP overrides, CA trust store, masked secrets). Recommend closing #4170 as completed/duplicate of delivered #4201 (+#4443 for UI), citing the commits above. No backports: every supported YBA release line already contains the 2020 delivery.

## Test strategy

Not applicable for a close-as-delivered outcome. Existing regression coverage of the delivered surface: `managed/src/test/java/com/yugabyte/yw/common/EmailHelperTest.java` (43 `smtpData` references — SMTP data resolution, properties conversion, defaults fallback) and `managed/src/test/java/com/yugabyte/yw/controllers/CustomerControllerTest.java` (8 `smtpData` references — PUT /customers smtpData create/edit/reset), both exercised through public API entry points.

## Confidentiality check

Contains only public artifacts: public GitHub issue/commit references, public source paths/line numbers from the open-source `yugabyte/yugabyte-db` repo, and public docs. The Aha link and example emails quoted are from the public issue/commit messages themselves. No customer data, PII, internal hostnames, or secrets.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4170-era-HealthManager.java.txt`):

// Copyright (c) YugaByte, Inc.

package com.yugabyte.yw.common;

import com.yugabyte.yw.models.Provider;

import com.google.inject.Inject;
import com.google.inject.Singleton;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.List;

import play.libs.Json;

@Singleton
public class HealthManager extends DevopsBase {
  @Inject
  play.Configuration appConfig;

  public static final String HEALTH_CHECK_SCRIPT = "bin/cluster_health.py";

  // TODO: we don't need this?
  private static final String YB_CLOUD_COMMAND_TYPE = "health_check";

  public static class ClusterInfo {
    public String identityFile = null;
    public int sshPort;
    // TODO: this is to be used by k8s.
    // Note: this is the same across all clusters, so maybe we should pull it out one level above.
    public Map<String, String> namespaceToConfig = new HashMap<>();
    public Map<String, String> masterNodes = new HashMap<>();
    public Map<String, String> tserverNodes = new HashMap<>();
    public String ybSoftwareVersion = null;
    public boolean enableTlsClient = false;
    public boolean enableYSQL = false;
    public int ysqlPort = 0;
  }

  public ShellProcessHandler.ShellResponse runCommand(
      Provider provider,
      List<ClusterInfo> clusters,
      String universeName,
      String customerTag,
      String destination,
      Long potentialStartTimeMs,
      Boolean sendMailAlways,
      Boolean reportOnlyErrors) {
    List<String> commandArgs = new ArrayList<>();

    commandArgs.add(PY_WRAPPER);
    commandArgs.add(HEALTH_CHECK_SCRIPT);
    commandArgs.add("--cluster_payload");
    commandArgs.add(Json.stringify(Json.toJson(clusters)));
    commandArgs.add("--universe_name");
    commandArgs.add(universeName);
    commandArgs.add("--customer_tag");
    commandArgs.add(customerTag);
    if (destination != null) {
      commandArgs.add("--destination");
      commandArgs.add(destination);
    }
    if (potentialStartTimeMs > 0) {
      commandArgs.add("--start_time_ms");
      commandArgs.add(String.valueOf(potentialStartTimeMs));
    }
    if (sendMailAlways) {
      commandArgs.add("--send_status");
    }
    // Start with a copy of the cloud config env vars.
    HashMap extraEnvVars = new HashMap<>(provider.getConfig());
    String emailUsername = appConfig.getString("yb.health.ses_email_username");
    if (emailUsername != null) {
      extraEnvVars.put("YB_ALERTS_USERNAME", emailUsername);
    }
    String emailPassword = appConfig.getString("yb.health.ses_email_password");
    if (emailPassword != null) {
      extraEnvVars.put("YB_ALERTS_PASSWORD", emailPassword);
    }
    String email = appConfig.getString("yb.health.default_email");
    if (email != null) {
      extraEnvVars.put("YB_ALERTS_EMAIL", email);
    }
    if (reportOnlyErrors) {
      commandArgs.add("--report_only_errors");
    }

    LOG.info("Command to run: [" + String.join(" ", commandArgs) + "]");
    return shellProcessHandler.run(commandArgs, extraEnvVars, false /*logCmdOutput*/);
  }

  @Override
  protected String getCommandType() {
    return YB_CLOUD_COMMAND_TYPE;
  }
}

**Comment by kriteshranjan** [comment]
AI validation notes (`yb4170-era-HealthChecker.java.txt`):

/*
 * Copyright 2019 YugaByte, Inc. and Contributors
 *
 * Licensed under the Polyform Free Trial License 1.0.0 (the "License"); you
 * may not use this file except in compliance with the License. You
 * may obtain a copy of the License at
 *
 *     https://github.com/YugaByte/yugabyte-db/blob/master/licenses/POLYFORM-FREE-TRIAL-LICENSE-1.0.0.txt
 */

package com.yugabyte.yw.commissioner;

import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.Date;
import java.util.HashMap;
import java.util.stream.Collectors;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.google.inject.Inject;
import com.google.inject.Singleton;

import com.yugabyte.yw.common.ShellProcessHandler;
import com.yugabyte.yw.common.HealthManager;
import com.yugabyte.yw.common.PlacementInfoUtil;
import com.yugabyte.yw.commissioner.Common.CloudType;
import com.yugabyte.yw.forms.NodeInstanceFormData;
import com.yugabyte.yw.forms.UniverseDefinitionTaskParams;
import com.yugabyte.yw.forms.CustomerRegisterFormData.AlertingData;
import com.yugabyte.yw.models.AccessKey;
import com.yugabyte.yw.models.Customer;
import com.yugabyte.yw.models.CustomerTask;
import com.yugabyte.yw.models.CustomerConfig;
import com.yugabyte.yw.models.HealthCheck;
import com.yugabyte.yw.models.NodeInstance;
import com.yugabyte.yw.models.Provider;
import com.yugabyte.yw.models.Universe;
import com.yugabyte.yw.models.helpers.NodeDetails;
import com.yugabyte.yw.common.Util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import play.api.Play;
import play.Configuration;
import play.libs.Json;

import akka.actor.ActorSystem;
import com.google.common.annotations.VisibleForTesting;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.TimeUnit;
import play.Environment;
import scala.concurrent.ExecutionContext;
import scala.concurrent.duration.Duration;

import io.prometheus.client.Gauge;
import io.prometheus.client.CollectorRegistry;

import com.fasterxml.jackson.databind.JsonNode;

@Singleton
public class HealthChecker {
  public static final Logger LOG = LoggerFactory.getLogger(HealthChecker.class);

  public static final String kUnivMetricName = "yb_univ_health_status";
  public static final String kUnivUUIDLabel = "univ_uuid";
  public static final String kUnivNameLabel = "univ_name";
  public static final String kCheckLabel = "check_name";
  public static final String kNodeLabel = "node";

  play.Configuration config;

  private long HEALTH_CHECK_SLEEP_INTERVAL_MS = 0;

  private long STATUS_UPDATE_INTERVAL_MS = 0;

  // Last time we sent a status update email per customer.
  private Map<UUID, Long> lastStatusUpdateTimeMap = new HashMap<>();

  // Last time we actually ran the health check script per customer.
  private Map<UUID, Long> lastCheckTimeMap = new HashMap<>();

  // What will run the health checking script.
  HealthManager healthManager;

  private Gauge healthMetric = null;

  private AtomicBoolean running = new AtomicBoolean(false);

  private final ActorSystem actorSystem;

  private final ExecutionContext executionContext;

  private final Environment environment;

  private CollectorRegistry promRegistry;

  public HealthChecker(
      ActorSystem actorSystem,
      Configuration config,
      Environment environment,
      ExecutionContext executionContext,
      HealthManager healthManager,
      CollectorRegistry promRegistry) {
    this.actorSystem = actorSystem;
    this.config = config;
    this.environment = environment;
    this.executionContext = executionContext;
    this.healthManager = healthManager;
    this.promRegistry = promRegistry;

    this.initialize();
  }


  @Inject
  public HealthChecker(
      ActorSystem actorSystem,
      Configuration config,
      Environment environment,
      ExecutionContext executionContext,
      HealthManager healthManager) {
        this(actorSystem, config, environment, executionContext,
             healthManager, CollectorRegistry.defaultRegistry);
  }

  private void initialize() {
    LOG.info("Scheduling health checker every " + this.healthCheckIntervalMs() + " ms");
    this.actorSystem.scheduler().schedule(
      Duration.create(0, TimeUnit.MILLISECONDS), // initialDelay
      Duration.create(this.healthCheckIntervalMs(), TimeUnit.MILLISECONDS), // interval
      () -> scheduleRunner(),
      this.executionContext
    );

    try {
      healthMetric = Gauge.build(kUnivMetricName, "Boolean result of health checks").
                           labelNames(kUnivUUIDLabel, kUnivNameLabel, kNodeLabel, kCheckLabel).
                           register(this.promRegistry);
    } catch (IllegalArgumentException e) {
      LOG.warn("Failed to build prometheus gauge for name: " + kUnivMetricName);
    }

  }

  // The interval at which the checker will run.
  // Can be overridden per customer.
  private long healthCheckIntervalMs() {
    Long interval = config.getLong("yb.health.check_interval_ms");
    return interval == null ? 0 : interval;
  }

  // The interval at which to send a status update of all the current universes.
  // Can be overridden per customer.
  private long statusUpdateIntervalMs() {
    Long interval = config.getLong("yb.health.status_interval_ms");
    return interval == null ? 0 : interval;
  }

  private String ybAlertEmail() {
    return config.getString("yb.health.default_email");
  }

  private void processResults(Universe u, String response) {
    Boolean hasErrors = false;
    try {
      JsonNode healthJSON = Util.convertStringToJson(response);
      for (JsonNode entry : healthJSON.path("data")) {
        String nodeName = entry.path("node").asText();
        String checkName = entry.path("message").asText();
        Boolean checkResult = entry.path("has_error").asBoolean();
        hasErrors = checkResult || hasErrors;
        if (null == healthMetric)
          continue;

        Gauge.Child prometheusVal = healthMetric.labels(
          u.universeUUID.toString(),
          u.name,
          nodeName,
          checkName
        );
        prometheusVal.set(checkResult ? 1 : 0);
      }
      LOG.info("Health check for universe " + u.name +
               (hasErrors ? " reported errors." : "reported success."));
     } catch (Exception e) {
      LOG.warn("Failed to convert health check response to prometheus metrics " + e.getMessage());
    }
  }

  @VisibleForTesting
  void scheduleRunner() {
    if (running.get()) {
      LOG.info("Previous run still underway");
      return;
    }

    LOG.info("Running health checker");
    running.set(true);
    // TODO(bogdan): This will not be too DB friendly when we go multi-tenant.
    for (Customer c : Customer.getAll()) {
      try {
        checkCustomer(c);
      } catch (Exception ex) {
        LOG.error("Error running health check for customer " + c.uuid, ex);
      }
    }
    LOG.info("Completed running health checker.");
    running.set(false);
  }

  public void checkCustomer(Customer c) {
    // We need an alerting config to do work.
    CustomerConfig config = CustomerConfig.getAlertConfig(c.uuid);
    if (config == null) {
      LOG.info("Skipping customer " + c.uuid + " due to missing alerting config...");
      return;
    }
    AlertingData alertingData = Json.fromJson(config.data, AlertingData.class);
    long now = (new Date()).getTime();
    long checkIntervalMs = alertingData.checkIntervalMs <= 0
      ? healthCheckIntervalMs()
      : alertingData.checkIntervalMs;
    boolean shouldRunCheck = (now - checkIntervalMs) >
        lastCheckTimeMap.getOrDefault(c.uuid, 0l);
    long statusUpdateIntervalMs = alertingData.statusUpdateIntervalMs <= 0
      ? statusUpdateIntervalMs()
      : alertingData.statusUpdateIntervalMs;
    boolean shouldSendStatusUpdate = (now - statusUpdateIntervalMs) >
        lastStatusUpdateTimeMap.getOrDefault(c.uuid, 0l);
    // Always do a check if it's time for a status update OR if it's time for a check.
    if (shouldSendStatusUpdate || shouldRunCheck) {
      // Since we'll do a check, update this all the time.
      lastCheckTimeMap.put(c.uuid, now);
      if (shouldSendStatusUpdate) {
        lastStatusUpdateTimeMap.put(c.uuid, now);
      }
      checkAllUniverses(c, config, shouldSendStatusUpdate);
    }
  }

  public void checkAllUniverses(
      Customer c, CustomerConfig config, boolean shouldSendStatusUpdate) {
    // Process all of a customer's universes.
    for (Universe u : c.getUniverses()) {
      try {
        checkSingleUniverse(u, c, config, shouldSendStatusUpdate);
      } catch (Exception ex) {
        LOG.error("Error running health check for universe " + u.universeUUID, ex);
      }
     }
  }

  public void checkSingleUniverse(
      Universe u, Customer c, CustomerConfig config, boolean shouldSendStatusUpdate) {
    // Validate universe data and make sure nothing is in progress.
    UniverseDefinitionTaskParams details = u.getUniverseDetails();
    if (details == null) {
      LOG.warn("Skipping universe " + u.name + " due to invalid details json...");
      return;
    }
    if (details.updateInProgress) {
      LOG.warn("Skipping universe " + u.name + " due to task in progress...");
      return;
    }
    LOG.info("Doing health check for universe: " + u.name);
    Map<UUID, HealthManager.ClusterInfo> clusterMetadata = new HashMap<>();
    boolean invalidUniverseData = false;
    String providerCode = "";
    for (UniverseDefinitionTaskParams.Cluster cluster : details.clusters) {
      HealthManager.ClusterInfo info = new HealthManager.ClusterInfo();
      clusterMetadata.put(cluster.uuid, info);
      info.ybSoftwareVersion = cluster.userIntent.ybSoftwareVersion;
      info.enableYSQL = cluster.userIntent.enableYSQL;
      // Since health checker only uses CQLSH, we only care about the
      // client to node encryption flag.
      info.enableTlsClient = cluster.userIntent.enableClientToNodeEncrypt;

      Provider provider = Provider.get(UUID.fromString(cluster.userIntent.provider));
      if (provider == null) {
        LOG.warn("Skipping universe " + u.name + " due to invalid provider "
            + cluster.userIntent.provider);
        invalidUniverseData = true;
        break;
      }
      providerCode = provider.code;
      if (providerCode.equals(Common.CloudType.kubernetes.toString())) {
        info.namespaceToConfig = PlacementInfoUtil.getConfigPerNamespace(
            cluster.placementInfo, details.nodePrefix, provider);
      }

      // TODO(bogdan): We do not have access to the default port constant at this level, as it is
      // baked in devops...hardcode it for now.
      // Default to 54422.
      int sshPort = 54422;
      if (providerCode.equals(Common.CloudType.onprem.toString())) {
        // For onprem, technically, each node could have a different port, but let's pick one of
        // the nodes for now and go from there.
        // TODO(bogdan): Improve the onprem support in devops to have port per node.
        for (NodeDetails nd : details.nodeDetailsSet) {
          NodeInstance onpremNode = NodeInstance.getByName(nd.nodeName);
          if (onpremNode != null) {
            NodeInstanceFormData.NodeInstanceData onpremDetails = onpremNode.getDetails();
            if (onpremDetails != null) {
              sshPort = onpremDetails.sshPort;
              break;
            }
          }
        }
      }
      AccessKey accessKey = AccessKey.get(provider.uuid, cluster.userIntent.accessKeyCode);
      if (accessKey == null || accessKey.getKeyInfo() == null) {
        if (!providerCode.equals(CloudType.kubernetes.toString())) {
          LOG.warn("Skipping universe " + u.name + " due to invalid access key...");
          invalidUniverseData = true;
          break;
        }
      } else {
        info.identityFile = accessKey.getKeyInfo().privateKey;
        info.sshPort = sshPort;
      }
      if (info.enableYSQL) {
        for (NodeDetails nd : details.nodeDetailsSet) {
          if (nd.isYsqlServer) {
           info.ysqlPort = nd.ysqlServerRpcPort;
           break;
          }
        }
      }
    }
    // If any clusters were invalid, abort for this universe.
    if (invalidUniverseData) {
      return;
    }
    for (NodeDetails nd : details.nodeDetailsSet) {
      HealthManager.ClusterInfo info = clusterMetadata.get(nd.placementUuid);
      if (info == null) {
        invalidUniverseData = true;
        LOG.warn(String.format(
              "Universe %s has node %s with invalid placement %s", u.name, nd.nodeName,
              nd.placementUuid));
        break;
      }
      // TODO: we do not have a good way of marking the whole universe as k8s only.
      if (nd.isMaster) {
        info.masterNodes.put(nd.cloudInfo.private_ip, nd.nodeName);
      }
      if (nd.isTserver) {
        info.tserverNodes.put(nd.cloudInfo.private_ip, nd.nodeName);
      }
    }
    // If any nodes were invalid, abort for this universe.
    if (invalidUniverseData) {
      return;
    }
    List<String> destinations = new ArrayList<String>();
    AlertingData alertingData = null;
    if (config != null) {
      alertingData = Json.fromJson(config.data, AlertingData.class);
      String ybEmail = ybAlertEmail();
      if (alertingData.sendAlertsToYb && ybEmail != null && !ybEmail.isEmpty()) {
        destinations.add(ybEmail);
      }
      if (alertingData.alertingEmail != null && !alertingData.alertingEmail.isEmpty()) {
        destinations.add(alertingData.alertingEmail);
      }
    }
    CustomerTask lastTask = CustomerTask.getLatestByUniverseUuid(u.universeUUID);
    long potentialStartTime = 0;
    if (lastTask != null && lastTask.getCompletionTime()!= null) {
      potentialStartTime = lastTask.getCompletionTime().getTime();
    }
    // If last check had errors, set the flag to send an email. If this check will have an error,
    // we would send an email anyway, but if this check shows a healthy universe, let's send an
    // email about it.
    HealthCheck lastCheck = HealthCheck.getLatest(u.universeUUID);
    boolean lastCheckHadErrors = lastCheck != null && lastCheck.hasError();
    // Setup customer tag including name and code, for ease of email parsing.
    String customerTag = String.format("[%s][%s]", c.name, c.code);
    Provider mainProvider = Provider.get(UUID.fromString(
          details.getPrimaryCluster().userIntent.provider));

    String disabledUntilStr = u.getConfig().getOrDefault(Universe.DISABLE_ALERTS_UNTIL, "0");
    long disabledUntilSecs = 0;
    try {
      disabledUntilSecs = Long.parseLong(disabledUntilStr);
    } catch (NumberFormatException ne) {
      LOG.warn("invalid universe config for disabled alerts: [ " + disabledUntilStr + " ]");
      disabledUntilSecs = 0;
    }
    boolean silenceEmails = ((System.currentTimeMillis() / 1000) <= disabledUntilSecs);

    boolean reportOnlyErrors = !shouldSendStatusUpdate &&
                               null != alertingData &&
                               alertingData.reportOnlyErrors;
    boolean sendMailAlways = (shouldSendStatusUpdate || lastCheckHadErrors);
    // Call devops and process response.
    ShellProcessHandler.ShellResponse response = healthManager.runCommand(
        mainProvider,
        new ArrayList(clusterMetadata.values()),
        u.name,
        customerTag,
        (destinations.size() == 0 || silenceEmails) ? null : String.join(",", destinations),
        potentialStartTime,
        sendMailAlways,
        reportOnlyErrors
    );

    if (response.code == 0) {
      processResults(u, response.message);
      HealthCheck.addAndPrune(u.universeUUID, u.customerId, response.message);
    } else {
      LOG.error(String.format(
          "Health check script got error: code (%s), msg: ", response.code, response.message));
    }
  }
}

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu
